dali/operators/reader/numpy_reader_gpu_op.cc

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+
+#include "dali/pipeline/data/views.h"
+#include "dali/kernels/transpose/transpose_gpu.h"
+#include "dali/core/static_switch.h"
+#include "dali/operators/reader/numpy_reader_gpu_op.h"
+
+namespace dali {
+
+void NumpyReaderGPU::Prefetch() {
+  // We actually prepare the next batch
+  TimeRange tr("NumpyReaderGPU::Prefetch #" + to_string(curr_batch_producer_), TimeRange::kRed);
+  DataReader<GPUBackend, ImageFileWrapperGPU>::Prefetch();
+  auto &curr_batch = prefetched_batch_queue_[curr_batch_producer_];
+  auto &curr_tensor_list = prefetched_batch_tensors_[curr_batch_producer_];
+
+  // get shapes
+  for (auto &sample : curr_batch) {
+    thread_pool_.AddWork([&sample](int tid) {
+        sample->read_meta_f();
```

Comment:
you have a bug here. You are capturing `sample` by reference and immediately you override it with the next sample. You should iterate through indices, capture them by value and access the element inside the lambda:
```
for (int i = ...) {
  thread_pool.AddWork([i](int tid) {
    curr_batch[i]->read_meta_f();
  });
}
```
```

Reply:
Done